core/src/main/java/io/grpc/ManagedChannelProvider.java

```diff
@@ -70,7 +70,13 @@ public int compare(ManagedChannelProvider f1, ManagedChannelProvider f2) {
   @VisibleForTesting
   public static Iterable<ManagedChannelProvider> getCandidatesViaServiceLoader(
       ClassLoader classLoader) {
-    return ServiceLoader.load(ManagedChannelProvider.class, classLoader);
+    Iterable<ManagedChannelProvider> i
+        = ServiceLoader.load(ManagedChannelProvider.class, classLoader);
+    if (!i.iterator().hasNext()
+            && classLoader == ManagedChannelProvider.class.getClassLoader()) {
+      i = ServiceLoader.load(ManagedChannelProvider.class);
```

Comment:
A comment about SPI fly would be great here. Ditto elsewhere.

Reply:
OK i will add this.